### PR TITLE
CBL-221 Keep Puller in busy status longer

### PR DESCRIPTION
**Pull request by borrrden**
It needs to wait until all the unfinished incoming revisions are done.  These are interjected in a thread-suspicious way by the Inserter.

### CODE REVIEW

**Comment by borrrden** [comment]
I'll see if I can add a test for this.  I think it should be possible with the loopback connection tests.

File: Replicator/Pusher.cc [hunk 1]
```diff
@@ -637,6 +637,9 @@ namespace litecore { namespace repl {
     Worker::ActivityLevel Pusher::computeActivityLevel() const {
         ActivityLevel level;
         if (!connection()) {
+            // Does this need a similar guard to what Puller has?  It doesn't
+            // seem so since the Puller has stuff that happens even after the
+            // connection is closed, while the Pusher does not seem to.
```
Comment (snej) [inline Replicator/Pusher.cc]: I agree, I don't think Pusher needs a change.

File: Replicator/Puller.cc [hunk 2]
```diff
@@ -350,7 +350,7 @@ namespace litecore { namespace repl {
     Worker::ActivityLevel Puller::computeActivityLevel() const {
         ActivityLevel level;
         if (_fatalError || !connection()) {
-            level = kC4Stopped;
+            level = _unfinishedIncomingRevs > 0 ? kC4Busy : kC4Stopped;
```
Comment (snej) [inline Replicator/Puller.cc]: I like the logic, but I think it'd be clearer if done by moving the existing `_unfinishedIncomingRevs` test to the start of the `if` chain:

```c++
if (_unfinishedIncomingRevs > 0) {
    level = kBusy;
} else if (_fatalError || !connection()) {
    level = kStopped;
} else ...

**Comment by snej** [comment]
Oh hey, while you're fixing this, it'd be good to change the bad cast that triggered the bug into a `dynamic_cast<Replicator*>`.

**Comment by borrrden** [comment]
The above seems to be taken care of in another PR, should I put it in this one too?

**Comment by snej** [comment]
No need to; yeah, I started thinking about unsafe casts and null pointers and cleaned up the replicator code.

**Pull request merged**